### PR TITLE
vic-machine suggests valid networks

### LINKED ISSUE #1711: vic-machine: suggest valid datastore, networks and datacenter
Similar to compute resource suggestion, we also need suggestion for datacenter, network and datastore

### LINKED ISSUE #1800: vDS validation output error
```
$ ./vic-machine-linux create --container-network "VM Network" -target 10.118.67.200 -user root -p password -i vsanDatastore --volume-store vsanDatastore/
volooms:vsan                                                                                                                                                                                                         
INFO[2016-08-03T13:28:36-07:00] ### Installing VCH ####                      
INFO[2016-08-03T13:28:36-07:00] Generating certificate/key pair - private key in ./virtual-container-host-key.pem 
INFO[2016-08-03T13:28:37-07:00] Validating supplied configuration            
INFO[2016-08-03T13:28:37-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.66.151" 
INFO[2016-08-03T13:28:37-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.66.150" 
INFO[2016-08-03T13:28:37-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.67.208" 
INFO[2016-08-03T13:28:37-07:00] Firewall configuration OK on hosts:          
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.66.151"         
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.66.150"         
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.67.208"         
INFO[2016-08-03T13:28:37-07:00] License check OK on hosts:                   
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.66.151"         
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.66.150"         
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/10.118.67.208"         
INFO[2016-08-03T13:28:37-07:00] DRS check OK on:                             
INFO[2016-08-03T13:28:37-07:00]   "/dc1/host/cluster1/Resources"             
ERRO[2016-08-03T13:28:37-07:00] --------------------                         
ERRO[2016-08-03T13:28:37-07:00] An existing distributed port group must be specified for bridge network on vCenter: network 'virtual-container-host' not found 
ERRO[2016-08-03T13:28:37-07:00] Unable to check hosts in vDS for "virtual-container-host": ServerFaultCode: 
Error processing attribute "type" with value ""

while parsing MoRef for ManagedObject of type vmodl.ManagedObject
at line 2, column 259

while parsing property "obj" of static type ManagedObject

while parsing serialized DataObject of type vmodl.query.PropertyCollector.ObjectSpec
at line 2, column 248

while parsing property "objectSet" of static type ArrayOfObjectSpec

while parsing serialized DataObject of type vmodl.query.PropertyCollector.FilterSpec
at line 2, column 161

while parsing call information for method RetrieveProperties
at line 2, column 66

while parsing SOAP body
at line 2, column 60

while parsing SOAP envelope
at line 2, column 0

while parsing HTTP request for method retrieveContents
on object of type vmodl.query.PropertyCollector
at line 1, column 0 
ERRO[2016-08-03T13:28:37-07:00] "VM Network" is not a Distributed Port Group 
ERRO[2016-08-03T13:28:37-07:00] Unable to check hosts in vDS for "VM Network": ServerFaultCode:
```

```
$ ./vic-machine-linux create --bridge-network VM\ Network -target 10.118.67.200 -user root -p password -i vsanDatastore --volume-store vsanDatastore/volooms:vsan
INFO[2016-08-03T13:35:24-07:00] ### Installing VCH ####                      
INFO[2016-08-03T13:35:24-07:00] Generating certificate/key pair - private key in ./virtual-container-host-key.pem 
INFO[2016-08-03T13:35:24-07:00] Validating supplied configuration            
INFO[2016-08-03T13:35:24-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.66.151" 
INFO[2016-08-03T13:35:25-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.66.150" 
INFO[2016-08-03T13:35:25-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.118.67.208" 
INFO[2016-08-03T13:35:25-07:00] Firewall configuration OK on hosts:          
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.66.151"         
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.66.150"         
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.67.208"         
INFO[2016-08-03T13:35:25-07:00] License check OK on hosts:                   
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.66.151"         
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.66.150"         
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/10.118.67.208"         
INFO[2016-08-03T13:35:25-07:00] DRS check OK on:                             
INFO[2016-08-03T13:35:25-07:00]   "/dc1/host/cluster1/Resources"             
ERRO[2016-08-03T13:35:25-07:00] --------------------                         
ERRO[2016-08-03T13:35:25-07:00] the bridge network must not be shared with another network role - external also uses "VM Network" 
ERRO[2016-08-03T13:35:25-07:00] the bridge network must not be shared with another network role - client also uses "VM Network" 
ERRO[2016-08-03T13:35:25-07:00] the bridge network must not be shared with another network role - management also uses "VM Network" 
ERRO[2016-08-03T13:35:25-07:00] An existing distributed port group must be specified for bridge network on vCenter: "VM Network" is not a Distributed Port Group 
ERRO[2016-08-03T13:35:25-07:00] Unable to check hosts in vDS for "VM Network": ServerFaultCode: 
Error processing attribute "type" with value ""

while parsing MoRef for ManagedObject of type vmodl.ManagedObject
at line 2, column 259

while parsing property "obj" of static type ManagedObject

while parsing serialized DataObject of type vmodl.query.PropertyCollector.ObjectSpec
at line 2, column 248

while parsing property "objectSet" of static type ArrayOfObjectSpec

while parsing serialized DataObject of type vmodl.query.PropertyCollector.FilterSpec
at line 2, column 161

while parsing call information for method RetrieveProperties
at line 2, column 66

while parsing SOAP body
at line 2, column 60

while parsing SOAP envelope
at line 2, column 0

while parsing HTTP request for method retrieveContents
on object of type vmodl.query.PropertyCollector
at line 1, column 0 
ERRO[2016-08-03T13:35:25-07:00] Create cannot continue: configuration validation failed 
ERRO[2016-08-03T13:35:25-07:00] --------------------                         
ERRO[2016-08-03T13:35:25-07:00] vic-machine-linux failed: validation of configuration failed
```

on vC this happens when trying to deploy and not specifying a bridge network and when specifying a bridge network that isn't a DPG

### PR DESCRIPTION
**Pull request by andrewtchin**
Suggest valid networks when the user enters incorrect network. 
Regression tests passed locally.

Fixes #1800, partial fix #1711 

ESX:
```
⇒  bin/vic-machine-linux create --target 192.168.218.177 --name chin-vic-110 --user root --password password --image-datastore datastore1 --external-network "ext" --container-network cont1 --container-network bridge --bridge-network asdf                                                                                                                              
INFO[2016-08-10T13:38:06-07:00] ### Installing VCH ####                      
INFO[2016-08-10T13:38:06-07:00] Generating certificate/key pair - private key in ./chin-vic-110-key.pem 
INFO[2016-08-10T13:38:07-07:00] Validating supplied configuration            
INFO[2016-08-10T13:38:07-07:00] Suggesting valid networks for --external-network 
INFO[2016-08-10T13:38:07-07:00] Suggested values for --external-network:     
INFO[2016-08-10T13:38:07-07:00]   "VM Network"                               
INFO[2016-08-10T13:38:07-07:00]   "docker-appliance"                         
INFO[2016-08-10T13:38:07-07:00]   "network"                                  
INFO[2016-08-10T13:38:07-07:00]   "chin-vic-110"                             
INFO[2016-08-10T13:38:07-07:00] Suggesting valid networks for --client-network 
INFO[2016-08-10T13:38:07-07:00] Suggested values for --client-network:       
INFO[2016-08-10T13:38:07-07:00]   "VM Network"                               
INFO[2016-08-10T13:38:07-07:00]   "docker-appliance"                         
INFO[2016-08-10T13:38:07-07:00]   "network"                                  
INFO[2016-08-10T13:38:07-07:00]   "chin-vic-110"                             
INFO[2016-08-10T13:38:07-07:00] Suggesting valid networks for --management-network 
INFO[2016-08-10T13:38:07-07:00] Suggested values for --management-network:   
INFO[2016-08-10T13:38:07-07:00]   "VM Network"                               
INFO[2016-08-10T13:38:07-07:00]   "docker-appliance"                         
INFO[2016-08-10T13:38:07-07:00]   "network"                                  
INFO[2016-08-10T13:38:07-07:00]   "chin-vic-110"                             
INFO[2016-08-10T13:38:07-07:00] Suggesting valid networks for --container-network 
INFO[2016-08-10T13:38:07-07:00] Suggested values for --container-network:    
INFO[2016-08-10T13:38:07-07:00]   "VM Network"                               
INFO[2016-08-10T13:38:07-07:00]   "docker-appliance"                         
INFO[2016-08-10T13:38:07-07:00]   "network"                                  
INFO[2016-08-10T13:38:07-07:00]   "chin-vic-110"                             
INFO[2016-08-10T13:38:07-07:00] Firewall status: DISABLED on "/ha-datacenter/host/localhost.localdomain/localhost.localdomain" 
INFO[2016-08-10T13:38:07-07:00] Firewall configuration OK on hosts:          
INFO[2016-08-10T13:38:07-07:00]   "/ha-datacenter/host/localhost.localdomain/localhost.localdomain" 
WARN[2016-08-10T13:38:07-07:00] Evaluation license detected. VIC may not function if evaluation expires or insufficient license is later assigned. 
INFO[2016-08-10T13:38:07-07:00] License check OK                             
INFO[2016-08-10T13:38:07-07:00] DRS check SKIPPED - target is standalone host 
ERRO[2016-08-10T13:38:07-07:00] --------------------                         
ERRO[2016-08-10T13:38:07-07:00] Error checking network for --external-network: network 'ext' not found 
ERRO[2016-08-10T13:38:07-07:00] Error checking network for --client-network: network 'ext' not found 
ERRO[2016-08-10T13:38:07-07:00] Error checking network for --management-network: network 'ext' not found 
ERRO[2016-08-10T13:38:07-07:00] Error adding container network "cont1": network 'cont1' not found 
ERRO[2016-08-10T13:38:07-07:00] Cannot use reserved name "bridge" for container network 
ERRO[2016-08-10T13:38:07-07:00] Create cannot continue: configuration validation failed 
ERRO[2016-08-10T13:38:07-07:00] --------------------                         
ERRO[2016-08-10T13:38:07-07:00] vic-machine-linux failed: validation of configuration failed
```

vC:
```
⇒  bin/vic-machine-linux create --target=10.161.4.108 --user=Administrator@vsphere.local --image-datastore=datastore1 --password=password --force=true --bridge-network=aaabridge --compute-resource=/ha-datacenter/host/10.160.220.41/Resources --external-network=extasdf --name=VCH-test --container-network cont1 --container-network cont2 --container-network bridge
INFO[2016-08-10T13:36:26-07:00] ### Installing VCH ####                      
INFO[2016-08-10T13:36:26-07:00] Generating certificate/key pair - private key in ./VCH-test-key.pem 
INFO[2016-08-10T13:36:28-07:00] Validating supplied configuration            
INFO[2016-08-10T13:36:28-07:00] Suggesting valid networks for --external-network 
INFO[2016-08-10T13:36:28-07:00] Suggested values for --external-network:     
INFO[2016-08-10T13:36:28-07:00]   "VM Network"                               
INFO[2016-08-10T13:36:28-07:00]   "management"                               
INFO[2016-08-10T13:36:28-07:00]   "vm-network"                               
INFO[2016-08-10T13:36:29-07:00] Suggesting valid networks for --client-network 
INFO[2016-08-10T13:36:29-07:00] Suggested values for --client-network:       
INFO[2016-08-10T13:36:29-07:00]   "VM Network"                               
INFO[2016-08-10T13:36:29-07:00]   "management"                               
INFO[2016-08-10T13:36:29-07:00]   "vm-network"                               
INFO[2016-08-10T13:36:29-07:00] Suggesting valid networks for --management-network 
INFO[2016-08-10T13:36:29-07:00] Suggested values for --management-network:   
INFO[2016-08-10T13:36:29-07:00]   "VM Network"                               
INFO[2016-08-10T13:36:29-07:00]   "management"                               
INFO[2016-08-10T13:36:29-07:00]   "vm-network"                               
INFO[2016-08-10T13:36:29-07:00] Suggesting valid networks for --bridge-network 
INFO[2016-08-10T13:36:30-07:00] Suggested values for --bridge-network:       
INFO[2016-08-10T13:36:30-07:00]   "bridge"                                   
INFO[2016-08-10T13:36:30-07:00]   "management"                               
INFO[2016-08-10T13:36:30-07:00]   "vm-network"                               
INFO[2016-08-10T13:36:30-07:00] Suggesting valid networks for --container-network 
INFO[2016-08-10T13:36:30-07:00] Suggested values for --container-network:    
INFO[2016-08-10T13:36:30-07:00]   "VM Network"                               
INFO[2016-08-10T13:36:30-07:00]   "management"                               
INFO[2016-08-10T13:36:30-07:00]   "vm-network"                               
INFO[2016-08-10T13:36:31-07:00] Firewall status: DISABLED on "/ha-datacenter/host/10.160.220.41/10.160.220.41" 
INFO[2016-08-10T13:36:31-07:00] Firewall configuration OK on hosts:          
INFO[2016-08-10T13:36:31-07:00]   "/ha-datacenter/host/10.160.220.41/10.160.220.41" 
INFO[2016-08-10T13:36:31-07:00] License check OK on hosts:                   
INFO[2016-08-10T13:36:31-07:00]   "/ha-datacenter/host/10.160.220.41/10.160.220.41" 
INFO[2016-08-10T13:36:31-07:00] DRS check SKIPPED - target is standalone host 
ERRO[2016-08-10T13:36:31-07:00] --------------------                         
ERRO[2016-08-10T13:36:31-07:00] Error checking network for --external-network: network 'extasdf' not found 
ERRO[2016-08-10T13:36:31-07:00] Error checking network for --client-network: network 'extasdf' not found 
ERRO[2016-08-10T13:36:31-07:00] Error checking network for --management-network: network 'extasdf' not found 
ERRO[2016-08-10T13:36:31-07:00] An existing distributed port group must be specified for bridge network on vCenter: network 'aaabridge' not found 
ERRO[2016-08-10T13:36:31-07:00] Error adding container network "cont1": network 'cont1' not found 
ERRO[2016-08-10T13:36:31-07:00] Error adding container network "cont2": network 'cont2' not found 
ERRO[2016-08-10T13:36:31-07:00] Cannot use reserved name "bridge" for container network 
ERRO[2016-08-10T13:36:31-07:00] Create cannot continue: configuration validation failed 
ERRO[2016-08-10T13:36:31-07:00] --------------------                         
ERRO[2016-08-10T13:36:31-07:00] vic-machine-linux failed: validation of configuration failed

```

### CODE REVIEW

File: lib/install/validate/validator.go [hunk 1]
```diff
@@ -368,6 +369,18 @@ func (v *Validator) inventoryPathToCluster(path string) string {
 	return strings.Join(pElems[:4], "/")
 }
 
+// inventoryPathToNetworkName is a convenience method that returns the network name based on
+// a path e.g. /ha-datacenter/network/VM Network
+func (v *Validator) inventoryPathToNetworkName(path string) string {
```
Comment (emlin) [inline lib/install/validate/validator.go]: path.Base could do this.

File: lib/install/validate/validator.go [hunk 2]
```diff
@@ -368,6 +369,18 @@ func (v *Validator) inventoryPathToCluster(path string) string {
...
+	r := regexp.MustCompile("/ha-datacenter/network/")
```
Comment (emlin) [inline lib/install/validate/validator.go]: btw, do not assume this name, this is different in VC.

File: lib/install/validate/network.go [hunk 3]
```diff
@@ -347,3 +373,78 @@ func (v *Validator) checkVDSMembership(ctx context.Context, network types.Manage
 	}
 	return nil
 }
+
+// suggestNetwork suggests all networks (only shows DPG on vC)
```
Comment (emlin) [inline lib/install/validate/network.go]: So far, we still allow standard network used in VC, if it's not used for bridge network and container network.
So maybe your suggestion code can add one more flag, for if standard network is expected.

File: lib/install/validate/network.go [hunk 4]
```diff
@@ -347,3 +373,78 @@ func (v *Validator) checkVDSMembership(ctx context.Context, network types.Manage
...
+
+	var dvp mo.DistributedVirtualPortgroup
+
+	r := object.NewDistributedVirtualPortgroup(v.Session.Client.Client, ref)
+	if err := r.Properties(v.Context, r.Reference(), []string{"tag"}, &dvp); err != nil {
+		log.Errorf("Unable to check tags on %q: %s", ref, err)
+		return false
+	}
+	for _, t := range dvp.Tag {
+		if t.Key == "SYSTEM/DVS.UPLINKPG" {
```
Comment (emlin) [inline lib/install/validate/network.go]: would suggest to check contains UPLINKPG or not. 
Another option is to check if there is one constant in govmomi for it, so avoid any VC internal message used here.

File: lib/install/validate/network.go [hunk 5]
```diff
@@ -32,86 +32,103 @@ import (
 	"golang.org/x/net/context"
 )
 
-func (v *Validator) network(ctx context.Context, input *data.Data, conf *config.VirtualContainerHostConfigSpec) {
+func (v *Validator) addNetworkHelper(ctx context.Context, conf *config.VirtualContainerHostConfigSpec, netName, epName, contNetName string, def bool) error {
```
Comment (emlin) [inline lib/install/validate/network.go]: would prefer the old format actually, although with some duplication. This new format is difficult to understand. 😸

File: lib/install/validate/validator.go (hunk 1 above)
Comment (andrewtchin) [inline lib/install/validate/validator.go]: 👍

File: lib/install/validate/validator.go (hunk 2 above)
Comment (andrewtchin) [inline lib/install/validate/validator.go]: removed

File: lib/install/validate/network.go (hunk 4 above)
Comment (andrewtchin) [inline lib/install/validate/network.go]: fixed

**Comment by emlin** [comment]
lgtm

**Comment by andrewtchin** [comment]
self lgtm

**Pull request merged**